openingd/openingd.c

```diff
@@ -769,6 +771,454 @@ static u8 *funder_channel_start(struct state *state,
+
+		/* TODO: add BOLT reference when merged
+		 * - MUST ensure the `output_info`.`script` is a standard script
+		 */
+		if (!is_known_scripttype(remote_outputs[i]->script))
+			peer_failed(state->pps,
+				    &state->channel_id,
+				    "Peer sent non-standard output script.");
+
+	}
+
+	/** TODO: add BOLT reference when merged
+	* The receiving node:
+	* - if the total `input_info`.`satoshis` is less than the total `output_info`.`satoshis`
+	*   - MUST fail the channel.
+	*/
+	if (!amount_sat_sub(&funding_tx_sats, funding, other_outputs))
+		peer_failed(state->pps,
+			    &state->channel_id,
+			    "Total remote input satoshi less than output satoshis. change:%s inputs:%s",
+			    type_to_string(tmpctx, struct amount_sat,
+					   &other_outputs),
+			    type_to_string(tmpctx, struct amount_sat,
+					   &funding));
+
+	if (!amount_sat_greater_eq(funding_tx_sats, stated_funding_sats))
+		peer_failed(state->pps,
+			    &state->channel_id,
+			    "Input amounts won't afford "
+			    "funding amount (desired: %s, provided: %s).",
+			    type_to_string(tmpctx, struct amount_sat,
+					   &stated_funding_sats),
+			    type_to_string(tmpctx, struct amount_sat,
+					   &funding_tx_sats));
+
+	return true;
+}
+#endif
+
+static u8 *funder_finalize_channel_setup2(struct state *state,
+					  struct utxo **utxos,
+					  struct bitcoin_signature *sig,
+					  struct bitcoin_tx **tx)
+{
+#if EXPERIMENTAL_FEATURES
+	u8 *msg;
+	struct channel_id id_in;
+	const u8 *wscript;
+	char *err_reason;
+	struct input_info **local_ins, **remote_ins;
+	struct output_info **local_outs, **remote_outs;
+	struct bitcoin_tx_input **remotetx_ins;
+	struct amount_sat total_funding;
+	struct bitcoin_signature their_sig, our_sig;
+	struct amount_msat local_funding_msat;
+
+	size_t i, input_count;
+	struct bitcoin_tx *funding_tx, *remote_commit, *local_commit;
+	struct witness_stack **remote_witnesses;
+
+	/* Derive components, omitting the funding output */
+	local_ins = tal_arr(tmpctx, struct input_info *, 0);
+	local_outs = tal_arr(tmpctx, struct output_info *, 0);
+	derive_input_output_info(tmpctx, *tx, utxos, true,
+				 &local_ins, &local_outs);
+
+	/* Send them to the peer */
+	msg = towire_funding_compose(tmpctx, &state->channel_id,
+				     cast_const2(const struct input_info **, local_ins),
+				     cast_const2(const struct output_info **, local_outs));
+
+	sync_crypto_write(state->pps, take(msg));
+
+	peer_billboard(false,
+		       "Opening channel: funding_compose sent, "
+		       "waiting for funding_compose reply");
+
+	msg = opening_negotiate_msg(tmpctx, state, false);
+	if (!msg)
+		return NULL;
+
+	/* The next message is "funding_compose", which tells us the funding
+	 * inputs and outputs they've selected. */
+	if (!fromwire_funding_compose(state, msg, &id_in,
+				      &remote_ins,
+				      &remote_outs))
+		peer_failed(state->pps,
+			    &state->channel_id,
+			    "Parsing received funding_compose");
```

Comment:
Send tal_hex() of msg, since it might be they sent something completely unexpected...